### PR TITLE
nixos/invoiceplane: add nginx as a webserver option

### PR DESCRIPTION
**Pull request by b12f**
## Description of changes

This adds nginx as an option for `services.invoiceplane.webserver`. I've tested this on my own instance.

Getting the vhost to play nice with phpfpm was done by following this community post: https://community.invoiceplane.com/t/topic/2654

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by teutat3s** [comment]
@ofborg build nixosTests.invoiceplane

File: nixos/modules/services/web-apps/invoiceplane.nix [hunk 1]
```diff
@@ -222,11 +222,10 @@ in
         };
 
         options.webserver = mkOption {
-          type = types.enum [ "caddy" ];
+          type = types.enum [ "caddy" "nginx" ];
           default = "caddy";
           description = lib.mdDoc ''
-            Which webserver to use for virtual host management. Currently only
-            caddy is supported.
+            Which webserver to use for virtual host management. Currently caddy and nginx are supported.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/invoiceplane.nix:228]:
```suggestion
            Which webserver to use for virtual host management.
```

File: nixos/modules/services/web-apps/invoiceplane.nix [hunk 2]
```diff
@@ -222,11 +222,10 @@ in
         };
 
         options.webserver = mkOption {
-          type = types.enum [ "caddy" ];
+          type = types.enum [ "caddy" "nginx" ];
           default = "caddy";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/invoiceplane.nix:226]:
```suggestion
          default = "caddy";
          example = "nginx";
```

File: nixos/modules/services/web-apps/invoiceplane.nix [hunk 3]
```diff
@@ -354,5 +353,39 @@ in
     };
   })
 
+  (mkIf (cfg.webserver == "nginx") {
+    services.nginx = {
+      enable = true;
+      virtualHosts = mapAttrs' (hostName: cfg: (
+        nameValuePair "${hostName}" {
+          root = "${pkg hostName cfg}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/invoiceplane.nix:361]:
```suggestion
          root = pkg hostName cfg;
```

File: nixos/modules/services/web-apps/invoiceplane.nix [hunk 4]
```diff
@@ -354,5 +353,39 @@ in
     };
   })
 
+  (mkIf (cfg.webserver == "nginx") {
+    services.nginx = {
+      enable = true;
+      virtualHosts = mapAttrs' (hostName: cfg: (
+        nameValuePair "${hostName}" {
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/invoiceplane.nix:360]:
```suggestion
        nameValuePair hostName {
```

File: nixos/modules/services/web-apps/invoiceplane.nix [hunk 5]
```diff
@@ -354,5 +353,39 @@ in
...
+            "/setup".extraConfig = ''
+              rewrite ^(.*)$ http://${hostName}/ redirect;
+            '';
+
+            "~ .php$" = {
+              extraConfig = ''
+                fastcgi_split_path_info ^(.+\.php)(/.+)$;
+                fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
+                fastcgi_pass unix:${config.services.phpfpm.pools."invoiceplane-${hostName}".socket};
+                include ${pkgs.nginx}/conf/fastcgi_params;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/invoiceplane.nix:380]: I feel like this should use services.nginx.package

**Comment by b12f** [comment]
Good suggestions @SuperSandro2000, thanks for the review. I've applied your changes.

**Review by teutat3s — APPROVED** [review]

**Comment by teutat3s** [comment]
@ofborg build nixosTests.invoiceplane

**Review by yayayayaka — APPROVED** [review]
Diff looking good to me.

**Pull request merged**